File: minigit-5/ts/src/commands/status.ts
// status command - Show working tree status

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, normalizePath, sha1 } from '../utils';
import { readIndex } from '../index-file';
import { getHeadCommit, getCurrentBranch, isDetachedHead } from '../refs';
import { walkTree, getTreeFromTreeIsh, hashObject } from '../objects';

export function status(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse flags
  let shortFormat = false;
  let porcelain = false;

  for (const arg of args) {
    if (arg === '--short' || arg === '-s') {
      shortFormat = true;
    } else if (arg === '--porcelain') {
      porcelain = true;
      shortFormat = true;
    }
  }

  // Get current branch
  const branch = getCurrentBranch(repoRoot);

  // Get HEAD tree files
  const headFiles = new Map<string, { sha: string; mode: string }>();
  const headSha = getHeadCommit(repoRoot);
  if (headSha) {
    try {
      const treeSha = getTreeFromTreeIsh(headSha, repoRoot);
      walkTree(treeSha, '', repoRoot).forEach((value, key) => headFiles.set(key, value));
    } catch {
      // No tree
    }
  }

  // Get index files
  const index = readIndex(repoRoot);
  const indexFiles = new Map<string, { sha: string; mode: number }>();
  for (const entry of index.entries) {
    indexFiles.set(entry.path, { sha: entry.sha, mode: entry.mode });
  }

  // Get working tree files
  const workingFiles = new Set<string>();
  collectWorkingFiles(repoRoot, repoRoot, workingFiles);

  // Calculate status
  const stagedNew: string[] = [];
  const stagedModified: string[] = [];
  const stagedDeleted: string[] = [];
  const unstagedModified: string[] = [];
  const unstagedDeleted: string[] = [];
  const untracked: string[] = [];

  // Compare HEAD with index (staged changes)
  for (const [filePath, indexEntry] of indexFiles) {
    const headEntry = headFiles.get(filePath);
    if (!headEntry) {
      stagedNew.push(filePath);
    } else if (headEntry.sha !== indexEntry.sha) {
      stagedModified.push(filePath);
    }
  }

  // Files in HEAD but not in index = staged deleted
  for (const [filePath] of headFiles) {
    if (!indexFiles.has(filePath)) {
      stagedDeleted.push(filePath);
    }
  }

  // Compare index with working tree (unstaged changes)
  for (const [filePath, indexEntry] of indexFiles) {
    const fullPath = path.join(repoRoot, filePath);

    if (!fs.existsSync(fullPath)) {
      unstagedDeleted.push(filePath);
    } else {
      // Check if modified
      const stats = fs.lstatSync(fullPath);
      if (!stats.isDirectory()) {
        let content: Buffer;
        if (stats.isSymbolicLink()) {
          content = Buffer.from(fs.readlinkSync(fullPath));
        } else {
          content = fs.readFileSync(fullPath);
        }
        const workingSha = hashObject('blob', content);
        if (workingSha !== indexEntry.sha) {
          unstagedModified.push(filePath);
        }
      }
    }
  }

  // Untracked files
  for (const filePath of workingFiles) {
    if (!indexFiles.has(filePath)) {
      untracked.push(filePath);
    }
  }

  // Sort all arrays
  stagedNew.sort();
  stagedModified.sort();
  stagedDeleted.sort();
  unstagedModified.sort();
  unstagedDeleted.sort();
  untracked.sort();

  // Output
  if (shortFormat) {
    // Short format: XY filename
    for (const f of stagedNew) {
      const y = unstagedModified.includes(f) ? 'M' : unstagedDeleted.includes(f) ? 'D' : ' ';
      console.log(`A${y} ${f}`);
    }
    for (const f of stagedModified) {
      const y = unstagedModified.includes(f) ? 'M' : unstagedDeleted.includes(f) ? 'D' : ' ';
      console.log(`M${y} ${f}`);
    }
    for (const f of stagedDeleted) {
      console.log(`D  ${f}`);
    }
    for (const f of unstagedModified) {
      if (!stagedNew.includes(f) && !stagedModified.includes(f)) {
        console.log(` M ${f}`);
      }
    }
    for (const f of unstagedDeleted) {
      if (!stagedNew.includes(f) && !stagedModified.includes(f)) {
        console.log(` D ${f}`);
      }
    }
    for (const f of untracked) {
      console.log(`?? ${f}`);
    }
  } else {
    // Long format
    if (isDetachedHead(repoRoot)) {
      console.log(`HEAD detached at ${headSha?.slice(0, 7) || '(unknown)'}`);
    } else {
      console.log(`On branch ${branch || 'main'}`);
    }

    const hasStaged = stagedNew.length > 0 || stagedModified.length > 0 || stagedDeleted.length > 0;
    const hasUnstaged = unstagedModified.length > 0 || unstagedDeleted.length > 0;
    const hasUntracked = untracked.length > 0;

    if (hasStaged) {
      console.log('');
      console.log('Changes to be committed:');
      console.log('  (use "minigit restore --staged <file>..." to unstage)');
      console.log('');
      for (const f of stagedNew) {
        console.log(`\tnew file:   ${f}`);
      }
      for (const f of stagedModified) {
        console.log(`\tmodified:   ${f}`);
      }
      for (const f of stagedDeleted) {
        console.log(`\tdeleted:    ${f}`);
      }
    }

    if (hasUnstaged) {
      console.log('');
      console.log('Changes not staged for commit:');
      console.log('  (use "minigit add <file>..." to update what will be committed)');
      console.log('');
      for (const f of unstagedModified) {
        console.log(`\tmodified:   ${f}`);
      }
      for (const f of unstagedDeleted) {
        console.log(`\tdeleted:    ${f}`);
      }
    }

    if (hasUntracked) {
      console.log('');
      console.log('Untracked files:');
      console.log('  (use "minigit add <file>..." to include in what will be committed)');
      console.log('');
      for (const f of untracked) {
        console.log(`\t${f}`);
      }
    }

    if (!hasStaged && !hasUnstaged && !hasUntracked) {
      console.log('nothing to commit, working tree clean');
    }
  }

  return 0;
}

function collectWorkingFiles(dir: string, repoRoot: string, result: Set<string>): void {
  const entries = fs.readdirSync(dir, { withFileTypes: true });

  for (const entry of entries) {
    if (entry.name === '.minigit') continue;

    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      collectWorkingFiles(fullPath, repoRoot, result);
    } else {
      const relativePath = normalizePath(path.relative(repoRoot, fullPath));
      result.add(relativePath);
    }
  }
}

File: minigit-5/py/commands/status.py
"""status command - Show working tree status."""

import sys
import os
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, get_file_mode
from objects import Blob, Tree, Commit
from index import Index
from refs import resolve_head, get_current_branch, is_head_detached


def run(args: list[str]) -> int:
    """Show the working tree status."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse flags
    short_format = '--short' in args or '-s' in args
    porcelain = '--porcelain' in args

    index = Index.read(repo_root)
    head_sha = resolve_head(repo_root)

    # Get HEAD tree entries
    head_entries = {}
    if head_sha:
        commit = Commit.read(head_sha, repo_root)
        head_entries = get_tree_entries(repo_root, commit.tree_sha)

    # Get working tree files
    work_tree_files = get_work_tree_files(repo_root)

    # Categorize changes
    staged_new = []
    staged_modified = []
    staged_deleted = []
    unstaged_modified = []
    unstaged_deleted = []
    untracked = []

    # Check index vs HEAD (staged changes)
    for name, entry in index.entries.items():
        if name not in head_entries:
            staged_new.append(name)
        elif entry.sha != head_entries[name][0]:
            staged_modified.append(name)

    # Check HEAD entries removed from index
    for name in head_entries:
        if name not in index.entries:
            staged_deleted.append(name)

    # Check working tree vs index (unstaged changes)
    for name, entry in index.entries.items():
        file_path = repo_root / name
        if not file_path.exists():
            if name not in staged_deleted:
                unstaged_deleted.append(name)
        else:
            work_sha = get_file_sha(file_path)
            if work_sha != entry.sha:
                unstaged_modified.append(name)

    # Find untracked files
    tracked = set(index.entries.keys())
    for rel_path in work_tree_files:
        if rel_path not in tracked:
            untracked.append(rel_path)

    # Output
    if short_format or porcelain:
        return output_short(staged_new, staged_modified, staged_deleted,
                          unstaged_modified, unstaged_deleted, untracked)
    else:
        return output_long(repo_root, staged_new, staged_modified, staged_deleted,
                         unstaged_modified, unstaged_deleted, untracked)


def get_tree_entries(repo_root: Path, tree_sha: str, prefix: str = '') -> dict:
    """Recursively get all entries from a tree."""
    entries = {}
    tree = Tree.read(tree_sha, repo_root)

    for entry in tree.entries:
        full_path = f"{prefix}{entry.name}" if prefix else entry.name
        if entry.mode == 0o40000:
            # Subtree
            entries.update(get_tree_entries(repo_root, entry.sha, full_path + '/'))
        else:
            entries[full_path] = (entry.sha, entry.mode)

    return entries


def get_work_tree_files(repo_root: Path) -> list:
    """Get all files in working tree."""
    files = []
    for item in repo_root.rglob('*'):
        if item.is_file() and '.minigit' not in item.parts:
            try:
                rel_path = item.relative_to(repo_root)
                files.append(str(rel_path).replace(os.sep, '/'))
            except ValueError:
                pass
    return sorted(files)


def get_file_sha(file_path: Path) -> str:
    """Get SHA of a file."""
    if file_path.is_symlink():
        content = os.readlink(file_path).encode()
    else:
        content = file_path.read_bytes()
    blob = Blob(content)
    return blob.sha


def output_short(staged_new, staged_modified, staged_deleted,
                unstaged_modified, unstaged_deleted, untracked) -> int:
    """Output short format status."""
    for name in sorted(staged_new):
        x = 'A'
        y = ' '
        if name in unstaged_modified:
            y = 'M'
        print(f"{x}{y} {name}")

    for name in sorted(staged_modified):
        x = 'M'
        y = ' '
        if name in unstaged_modified:
            y = 'M'
        print(f"{x}{y} {name}")

    for name in sorted(staged_deleted):
        print(f"D  {name}")

    for name in sorted(unstaged_modified):
        if name not in staged_new and name not in staged_modified:
            print(f" M {name}")

    for name in sorted(unstaged_deleted):
        if name not in staged_deleted:
            print(f" D {name}")

    for name in sorted(untracked):
        print(f"?? {name}")

    return 0


def output_long(repo_root, staged_new, staged_modified, staged_deleted,
               unstaged_modified, unstaged_deleted, untracked) -> int:
    """Output long format status."""
    branch = get_current_branch(repo_root)
    if branch:
        print(f"On branch {branch}")
    else:
        print("HEAD detached")

    has_staged = staged_new or staged_modified or staged_deleted
    has_unstaged = unstaged_modified or unstaged_deleted

    if has_staged:
        print("\nChanges to be committed:")
        print("  (use \"minigit restore --staged <file>...\" to unstage)")
        for name in sorted(staged_new):
            print(f"\tnew file:   {name}")
        for name in sorted(staged_modified):
            print(f"\tmodified:   {name}")
        for name in sorted(staged_deleted):
            print(f"\tdeleted:    {name}")

    if has_unstaged:
        print("\nChanges not staged for commit:")
        print("  (use \"minigit add <file>...\" to update what will be committed)")
        for name in sorted(unstaged_modified):
            if name not in staged_new:
                print(f"\tmodified:   {name}")
        for name in sorted(unstaged_deleted):
            if name not in staged_deleted:
                print(f"\tdeleted:    {name}")

    if untracked:
        print("\nUntracked files:")
        print("  (use \"minigit add <file>...\" to include in what will be committed)")
        for name in sorted(untracked):
            print(f"\t{name}")

    if not has_staged and not has_unstaged and not untracked:
        print("\nnothing to commit, working tree clean")

    return 0
